File: scripts/migrate-originals-to-archive.mjs
#!/usr/bin/env node
/**
 * Migration Script: Move originals from hot bucket to archive bucket
 *
 * This script:
 * 1. Queries the database for assets with originals in the wrong bucket
 * 2. Copies each file to the archive bucket
 * 3. Verifies the copy (size match)
 * 4. Updates the database
 * 5. Deletes from source (optional, disabled by default)
 *
 * Usage:
 *   node migrate-originals-to-archive.mjs [--dry-run] [--delete-source] [--batch-size=100]
 *
 * Environment variables required:
 *   DATABASE_URL or DB_URL
 *   STORAGE_S3_ACCESS_KEY_ID, STORAGE_S3_SECRET_ACCESS_KEY (hot bucket)
 *   STORAGE_S3_ARCHIVE_ACCESS_KEY_ID, STORAGE_S3_ARCHIVE_SECRET_ACCESS_KEY (archive bucket)
 *   STORAGE_S3_ENDPOINT (hot bucket endpoint)
 *   STORAGE_S3_ARCHIVE_ENDPOINT (archive bucket endpoint)
 *   STORAGE_S3_HOT_BUCKET
 *   STORAGE_S3_ARCHIVE_BUCKET
 */

import { S3Client, CopyObjectCommand, HeadObjectCommand, DeleteObjectCommand } from '@aws-sdk/client-s3';
import pg from 'pg';

// Parse command line arguments
const args = process.argv.slice(2);
const DRY_RUN = args.includes('--dry-run');
const DELETE_SOURCE = args.includes('--delete-source');
const BATCH_SIZE = parseInt(args.find(a => a.startsWith('--batch-size='))?.split('=')[1] || '100');

// Configuration from environment
const config = {
  database: process.env.DATABASE_URL || process.env.DB_URL,
  hotBucket: {
    name: process.env.STORAGE_S3_HOT_BUCKET || 'user-media',
    endpoint: process.env.STORAGE_S3_ENDPOINT || 'https://t3.storage.dev',
    accessKeyId: process.env.STORAGE_S3_ACCESS_KEY_ID,
    secretAccessKey: process.env.STORAGE_S3_SECRET_ACCESS_KEY,
  },
  archiveBucket: {
    name: process.env.STORAGE_S3_ARCHIVE_BUCKET || 'user-media-ira',
    endpoint: process.env.STORAGE_S3_ARCHIVE_ENDPOINT || 'https://fly.storage.tigris.dev',
    accessKeyId: process.env.STORAGE_S3_ARCHIVE_ACCESS_KEY_ID,
    secretAccessKey: process.env.STORAGE_S3_ARCHIVE_SECRET_ACCESS_KEY,
  },
};

// Validate configuration
function validateConfig() {
  const missing = [];
  if (!config.database) missing.push('DATABASE_URL or DB_URL');
  if (!config.hotBucket.accessKeyId) missing.push('STORAGE_S3_ACCESS_KEY_ID');
  if (!config.hotBucket.secretAccessKey) missing.push('STORAGE_S3_SECRET_ACCESS_KEY');
  if (!config.archiveBucket.accessKeyId) missing.push('STORAGE_S3_ARCHIVE_ACCESS_KEY_ID');
  if (!config.archiveBucket.secretAccessKey) missing.push('STORAGE_S3_ARCHIVE_SECRET_ACCESS_KEY');

  if (missing.length > 0) {
    console.error('Missing required environment variables:', missing.join(', '));
    process.exit(1);
  }
}

// Create S3 clients
function createS3Client(bucketConfig) {
  return new S3Client({
    endpoint: bucketConfig.endpoint,
    region: 'auto',
    credentials: {
      accessKeyId: bucketConfig.accessKeyId,
      secretAccessKey: bucketConfig.secretAccessKey,
    },
    forcePathStyle: true,
  });
}

// Get object metadata (size)
async function getObjectSize(client, bucket, key) {
  try {
    const response = await client.send(new HeadObjectCommand({ Bucket: bucket, Key: key }));
    return response.ContentLength;
  } catch (error) {
    if (error.name === 'NotFound' || error.$metadata?.httpStatusCode === 404) {
      return null;
    }
    throw error;
  }
}

// Copy object between buckets (cross-account copy via download/upload)
async function copyObject(sourceClient, destClient, sourceBucket, destBucket, key) {
  // For cross-account/cross-endpoint copies, we need to download and re-upload
  // Using the S3 copy command won't work across different endpoints

  const { GetObjectCommand, PutObjectCommand } = await import('@aws-sdk/client-s3');

  // Get the object from source
  const getResponse = await sourceClient.send(new GetObjectCommand({
    Bucket: sourceBucket,
    Key: key,
  }));

  // Upload to destination
  await destClient.send(new PutObjectCommand({
    Bucket: destBucket,
    Key: key,
    Body: getResponse.Body,
    ContentType: getResponse.ContentType,
    ContentLength: getResponse.ContentLength,
  }));

  return getResponse.ContentLength;
}

// Delete object
async function deleteObject(client, bucket, key) {
  await client.send(new DeleteObjectCommand({ Bucket: bucket, Key: key }));
}

// Main migration function
async function migrate() {
  validateConfig();

  console.log('=== Original Files Migration Script ===');
  console.log(`Mode: ${DRY_RUN ? 'DRY RUN (no changes)' : 'LIVE'}`);
  console.log(`Delete source after copy: ${DELETE_SOURCE}`);
  console.log(`Batch size: ${BATCH_SIZE}`);
  console.log(`Hot bucket: ${config.hotBucket.name} @ ${config.hotBucket.endpoint}`);
  console.log(`Archive bucket: ${config.archiveBucket.name} @ ${config.archiveBucket.endpoint}`);
  console.log('');

  // Connect to database
  const pool = new pg.Pool({ connectionString: config.database });
  const hotClient = createS3Client(config.hotBucket);
  const archiveClient = createS3Client(config.archiveBucket);

  try {
    // Count total assets to migrate
    const countResult = await pool.query(`
      SELECT COUNT(*) as count
      FROM asset
      WHERE "storageBackend" = 's3'
        AND "s3Bucket" = $1
        AND "s3Key" IS NOT NULL
    `, [config.hotBucket.name]);

    const totalCount = parseInt(countResult.rows[0].count);
    console.log(`Found ${totalCount} assets to migrate`);

    if (totalCount === 0) {
      console.log('Nothing to migrate!');
      return;
    }

    let processed = 0;
    let succeeded = 0;
    let skipped = 0;
    let failed = 0;

    // Process in batches
    while (processed < totalCount) {
      const batchResult = await pool.query(`
        SELECT id, "s3Key", "s3Bucket"
        FROM asset
        WHERE "storageBackend" = 's3'
          AND "s3Bucket" = $1
          AND "s3Key" IS NOT NULL
        ORDER BY "createdAt"
        LIMIT $2
        OFFSET $3
      `, [config.hotBucket.name, BATCH_SIZE, processed]);

      if (batchResult.rows.length === 0) break;

      for (const asset of batchResult.rows) {
        processed++;
        const key = asset.s3Key;

        try {
          // Check if already exists in archive (for resumability)
          const archiveSize = await getObjectSize(archiveClient, config.archiveBucket.name, key);

          if (archiveSize !== null) {
            // Already in archive, just update DB if needed
            if (asset.s3Bucket !== config.archiveBucket.name) {
              if (!DRY_RUN) {
                await pool.query(`
                  UPDATE asset SET "s3Bucket" = $1 WHERE id = $2
                `, [config.archiveBucket.name, asset.id]);
              }
              console.log(`[${processed}/${totalCount}] ${key} - already in archive, updated DB`);
            } else {
              console.log(`[${processed}/${totalCount}] ${key} - skipped (already migrated)`);
            }
            skipped++;
            continue;
          }

          // Get source size
          const sourceSize = await getObjectSize(hotClient, config.hotBucket.name, key);
          if (sourceSize === null) {
            console.error(`[${processed}/${totalCount}] ${key} - ERROR: not found in source bucket`);
            failed++;
            continue;
          }

          if (DRY_RUN) {
            console.log(`[${processed}/${totalCount}] ${key} - would copy (${(sourceSize / 1024 / 1024).toFixed(2)} MB)`);
            succeeded++;
            continue;
          }

          // Copy to archive
          console.log(`[${processed}/${totalCount}] ${key} - copying (${(sourceSize / 1024 / 1024).toFixed(2)} MB)...`);
          await copyObject(hotClient, archiveClient, config.hotBucket.name, config.archiveBucket.name, key);

          // Verify copy
          const destSize = await getObjectSize(archiveClient, config.archiveBucket.name, key);
          if (destSize !== sourceSize) {
            console.error(`[${processed}/${totalCount}] ${key} - ERROR: size mismatch (source=${sourceSize}, dest=${destSize})`);
            failed++;
            continue;
          }

          // Update database
          await pool.query(`
            UPDATE asset SET "s3Bucket" = $1 WHERE id = $2
          `, [config.archiveBucket.name, asset.id]);

          // Delete from source (optional)
          if (DELETE_SOURCE) {
            await deleteObject(hotClient, config.hotBucket.name, key);
            console.log(`[${processed}/${totalCount}] ${key} - migrated and deleted from source`);
          } else {
            console.log(`[${processed}/${totalCount}] ${key} - migrated (source retained)`);
          }

          succeeded++;

        } catch (error) {
          console.error(`[${processed}/${totalCount}] ${key} - ERROR: ${error.message}`);
          failed++;
        }
      }

      // Progress summary every batch
      console.log(`\n--- Progress: ${processed}/${totalCount} (${succeeded} succeeded, ${skipped} skipped, ${failed} failed) ---\n`);
    }

    // Final summary
    console.log('\n=== Migration Complete ===');
    console.log(`Total processed: ${processed}`);
    console.log(`Succeeded: ${succeeded}`);
    console.log(`Skipped (already migrated): ${skipped}`);
    console.log(`Failed: ${failed}`);

  } finally {
    await pool.end();
  }
}

// Run
migrate().catch(error => {
  console.error('Migration failed:', error);
  process.exit(1);
});

File: scripts/migrate-originals-to-archive.py
#!/usr/bin/env python3
"""
Migration Script: Move originals from hot bucket to archive bucket

Usage:
  python migrate.py --dry-run
  python migrate.py
  python migrate.py --delete-source
  python migrate.py --concurrency=10
  python migrate.py --limit=100
"""

import os
import sys
import asyncio
import argparse
from datetime import datetime

import boto3
import psycopg
from botocore.config import Config

# Parse arguments
parser = argparse.ArgumentParser()
parser.add_argument('--dry-run', action='store_true')
parser.add_argument('--delete-source', action='store_true')
parser.add_argument('--concurrency', type=int, default=10)
parser.add_argument('--batch-size', type=int, default=100)
parser.add_argument('--limit', type=int, default=0)
args = parser.parse_args()

# Config from environment
config = {
    'database': os.environ.get('DATABASE_URL') or os.environ.get('DB_URL'),
    'hot': {
        'bucket': os.environ.get('STORAGE_S3_HOT_BUCKET', 'user-media'),
        'endpoint': os.environ.get('STORAGE_S3_ENDPOINT', 'https://t3.storage.dev'),
        'access_key': os.environ.get('STORAGE_S3_ACCESS_KEY_ID'),
        'secret_key': os.environ.get('STORAGE_S3_SECRET_ACCESS_KEY'),
    },
    'archive': {
        'bucket': os.environ.get('STORAGE_S3_ARCHIVE_BUCKET', 'user-media-ira'),
        'endpoint': os.environ.get('STORAGE_S3_ARCHIVE_ENDPOINT', 'https://fly.storage.tigris.dev'),
        'access_key': os.environ.get('STORAGE_S3_ARCHIVE_ACCESS_KEY_ID'),
        'secret_key': os.environ.get('STORAGE_S3_ARCHIVE_SECRET_ACCESS_KEY'),
    },
}

# Stats
stats = {'processed': 0, 'succeeded': 0, 'skipped': 0, 'failed': 0, 'bytes': 0}
stats_lock = asyncio.Lock()
start_time = datetime.now()

def create_s3_client(cfg):
    return boto3.client(
        's3',
        endpoint_url=cfg['endpoint'],
        aws_access_key_id=cfg['access_key'],
        aws_secret_access_key=cfg['secret_key'],
        region_name='auto',
        config=Config(signature_version='s3v4')
    )

def get_object_size(client, bucket, key):
    try:
        response = client.head_object(Bucket=bucket, Key=key)
        return response['ContentLength']
    except client.exceptions.ClientError as e:
        if e.response['Error']['Code'] == '404':
            return None
        raise

def copy_object_streaming(hot_client, archive_client, hot_bucket, archive_bucket, key):
    """Stream copy without loading entire file into memory"""
    # Get object with streaming body
    response = hot_client.get_object(Bucket=hot_bucket, Key=key)
    content_type = response.get('ContentType', 'application/octet-stream')
    content_length = response['ContentLength']

    # Upload with streaming body
    archive_client.upload_fileobj(
        response['Body'],
        archive_bucket,
        key,
        ExtraArgs={'ContentType': content_type}
    )

    return content_length

async def process_asset(asset, hot_client, archive_client, db_pool, total_count, semaphore):
    async with semaphore:
        asset_id, key = asset['id'], asset['s3Key']

        async with stats_lock:
            stats['processed'] += 1
            idx = stats['processed']

        try:
            # Run S3 operations in thread pool (boto3 is sync)
            loop = asyncio.get_event_loop()

            # Check source exists
            source_size = await loop.run_in_executor(
                None, get_object_size, hot_client, config['hot']['bucket'], key
            )
            if source_size is None:
                print(f"[{idx}/{total_count}] {key} - ERROR: not found in source")
                async with stats_lock:
                    stats['failed'] += 1
                return

            # Check if already in archive
            archive_size = await loop.run_in_executor(
                None, get_object_size, archive_client, config['archive']['bucket'], key
            )

            if archive_size is not None and archive_size == source_size:
                # Already migrated, just update DB
                if not args.dry_run:
                    async with db_pool.connection() as conn:
                        await conn.execute(
                            'UPDATE asset SET "s3Bucket" = %s WHERE id = %s',
                            (config['archive']['bucket'], asset_id)
                        )
                print(f"[{idx}/{total_count}] {key} - already in archive, updated DB")
                async with stats_lock:
                    stats['skipped'] += 1
                return

            size_mb = source_size / 1024 / 1024

            if args.dry_run:
                print(f"[{idx}/{total_count}] {key} - would copy ({size_mb:.2f} MB)")
                async with stats_lock:
                    stats['succeeded'] += 1
                    stats['bytes'] += source_size
                return

            # Copy with streaming
            t0 = datetime.now()
            await loop.run_in_executor(
                None, copy_object_streaming,
                hot_client, archive_client,
                config['hot']['bucket'], config['archive']['bucket'], key
            )
            duration = (datetime.now() - t0).total_seconds()

            # Verify
            dest_size = await loop.run_in_executor(
                None, get_object_size, archive_client, config['archive']['bucket'], key
            )
            if dest_size != source_size:
                print(f"[{idx}/{total_count}] {key} - ERROR: size mismatch")
                async with stats_lock:
                    stats['failed'] += 1
                return

            # Update DB
            async with db_pool.connection() as conn:
                await conn.execute(
                    'UPDATE asset SET "s3Bucket" = %s WHERE id = %s',
                    (config['archive']['bucket'], asset_id)
                )

            # Delete source if requested
            if args.delete_source:
                await loop.run_in_executor(
                    None, hot_client.delete_object,
                    {'Bucket': config['hot']['bucket'], 'Key': key}
                )
                print(f"[{idx}/{total_count}] {key} - {size_mb:.2f} MB in {duration:.1f}s (deleted)")
            else:
                print(f"[{idx}/{total_count}] {key} - {size_mb:.2f} MB in {duration:.1f}s")

            async with stats_lock:
                stats['succeeded'] += 1
                stats['bytes'] += source_size

        except Exception as e:
            print(f"[{idx}/{total_count}] {key} - ERROR: {e}")
            async with stats_lock:
                stats['failed'] += 1

async def main():
    print("=== Original Files Migration Script (Python) ===")
    print(f"Mode: {'DRY RUN' if args.dry_run else 'LIVE'}")
    print(f"Delete source: {args.delete_source}")
    print(f"Concurrency: {args.concurrency}")
    print(f"Batch size: {args.batch_size}")
    if args.limit:
        print(f"Limit: {args.limit}")
    print(f"Hot bucket: {config['hot']['bucket']} @ {config['hot']['endpoint']}")
    print(f"Archive bucket: {config['archive']['bucket']} @ {config['archive']['endpoint']}")
    print()

    # Create clients
    hot_client = create_s3_client(config['hot'])
    archive_client = create_s3_client(config['archive'])

    # Database pool
    db_pool = psycopg.AsyncConnectionPool(config['database'], min_size=1, max_size=args.concurrency + 5)
    await db_pool.open()

    try:
        # Count assets
        async with db_pool.connection() as conn:
            result = await conn.execute(
                '''SELECT COUNT(*) FROM asset
                   WHERE "storageBackend" = 's3'
                   AND "s3Bucket" = %s
                   AND "s3Key" IS NOT NULL''',
                (config['hot']['bucket'],)
            )
            row = await result.fetchone()
            total_in_db = row[0]

        total_count = min(total_in_db, args.limit) if args.limit else total_in_db
        print(f"Found {total_in_db} assets, processing {total_count}\n")

        if total_count == 0:
            print("Nothing to migrate!")
            return

        semaphore = asyncio.Semaphore(args.concurrency)
        last_created_at = None
        last_id = None

        while stats['processed'] < total_count:
            remaining = min(args.batch_size, total_count - stats['processed']) if args.limit else args.batch_size

            async with db_pool.connection() as conn:
                if last_created_at is None:
                    result = await conn.execute(
                        '''SELECT id, "s3Key", "createdAt" FROM asset
                           WHERE "storageBackend" = 's3'
                           AND "s3Bucket" = %s
                           AND "s3Key" IS NOT NULL
                           ORDER BY "createdAt", id
                           LIMIT %s''',
                        (config['hot']['bucket'], remaining)
                    )
                else:
                    result = await conn.execute(
                        '''SELECT id, "s3Key", "createdAt" FROM asset
                           WHERE "storageBackend" = 's3'
                           AND "s3Bucket" = %s
                           AND "s3Key" IS NOT NULL
                           AND ("createdAt" > %s OR ("createdAt" = %s AND id > %s))
                           ORDER BY "createdAt", id
                           LIMIT %s''',
                        (config['hot']['bucket'], last_created_at, last_created_at, last_id, remaining)
                    )

                rows = await result.fetchall()

            if not rows:
                break

            assets = [{'id': r[0], 's3Key': r[1], 'createdAt': r[2]} for r in rows]
            last_created_at = assets[-1]['createdAt']
            last_id = assets[-1]['id']

            # Process batch concurrently
            tasks = [process_asset(a, hot_client, archive_client, db_pool, total_count, semaphore) for a in assets]
            await asyncio.gather(*tasks)

            # Progress
            elapsed = (datetime.now() - start_time).total_seconds()
            rate = stats['succeeded'] / elapsed if elapsed > 0 else 0
            gb = stats['bytes'] / 1024 / 1024 / 1024
            print(f"\n--- {stats['processed']}/{total_count} | OK: {stats['succeeded']} | Skip: {stats['skipped']} | Fail: {stats['failed']} | {gb:.2f} GB | {rate:.1f}/s | {elapsed:.0f}s ---\n")

        # Final summary
        elapsed = (datetime.now() - start_time).total_seconds()
        gb = stats['bytes'] / 1024 / 1024 / 1024
        print("\n=== Migration Complete ===")
        print(f"Processed: {stats['processed']}")
        print(f"Succeeded: {stats['succeeded']}")
        print(f"Skipped: {stats['skipped']}")
        print(f"Failed: {stats['failed']}")
        print(f"Data: {gb:.2f} GB")
        print(f"Time: {elapsed:.0f}s")

    finally:
        await db_pool.close()

if __name__ == '__main__':
    asyncio.run(main())
